predicators/approaches/delivery_approach.py

```diff
@@ -0,0 +1,197 @@
+"""An approach that implements a delivery-specific policy.
+
+Example command line:
+    python predicators/main.py --approach delivery_policy --seed 0 \
+        --env pddl_easy_delivery_procedural_tasks
+"""
+
+from typing import Callable, Iterable, Optional, Set, cast
+
+import numpy as np
+
+from predicators.approaches import ApproachFailure, BaseApproach
+from predicators.envs.pddl_env import _PDDLEnvState
+from predicators.structs import Action, GroundAtom, Object, Predicate, State, \
+    Task, _Option
+
+
+def filter_predicated_objs(objs: Iterable[Object], pred: Predicate,
+                           ground_atoms: Iterable[Object]) -> Set[GroundAtom]:
+    """Filters the objects based on whether they fulfill a unitary predicate in
+    a given set of ground atoms.
+
+    The predicate has to be unitary
+    """
+    return set(
+        filter(lambda obj: GroundAtom(pred, [obj]) in ground_atoms, objs))
+
+
+def check_option(ground_option: _Option, state: _PDDLEnvState) -> None:
+    if not ground_option.initiable(state):
+        raise ApproachFailure("Could not execute option")
+
+
+class DeliverySpecificApproach(BaseApproach):
+    """Implements a policy for the delivery domain.
+
+    See envs/assets/pddl/delivery/domain.pddl for the domain definition
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # Extracts domain-specific class instances for the delivery domain
+        types = {t.name: t for t in self._types}
+        self.type_loc = types["loc"]
+        self.type_paper = types["paper"]
+
+        options = {o.name: o for o in self._initial_options}
+        self.opt_pick_up = options["pick-up"]
+        self.opt_move = options["move"]
+        self.opt_deliver = options["deliver"]
+
+        predicates = {p.name: p for p in self._initial_predicates}
+        self.pred_is_home_base = predicates["ishomebase"]
+        self.pred_safe = predicates["safe"]
+        self.pred_at = predicates["at"]
+        self.pred_wants_paper = predicates["wantspaper"]
+        self.pred_satisfied = predicates["satisfied"]
+        self.pred_unpacked = predicates["unpacked"]
+        self.pred_carrying = predicates["carrying"]
+
+    @classmethod
+    def get_name(cls) -> str:
+        return "delivery_policy"
+
+    @property
+    def is_learning_based(self) -> bool:
+        return False
+
+    def _extract_at_loc_optional(
```

Comment:
I think this logic could be included inside `_extract_at_loc`, but that's just my personal preference :)

Reply:
I separated it mostly for the sake of typing

Reply:
if you assert something isn't None, mypy will be smart enough to incorporate that

Reply:
yeah but I wanted a nice error message printed